app/services/subscriptions/free_trial_billing_service.rb

```diff
@@ -10,36 +10,52 @@ def initialize(timestamp: Time.current)
 
     def call
       ending_trial_subscriptions.each do |subscription|
-        # if subscription.plan_pay_in_advance
-        #   BillSubscriptionJob.perform_later([subscription], timestamp)
-        # end
+        if subscription.plan_pay_in_advance &&
+           !subscription.was_already_billed_today &&
+           !already_billed_on_day_one?(subscription)
+          BillSubscriptionJob.perform_later([subscription], timestamp, recurring: false, skip_charges: true)
+        end
 
-        # TODO: Send webhook
-        subscription.update!(trial_ended_at: timestamp)
+        subscription.update!(trial_ended_at: subscription.trial_end_utc_date_from_query)
+
+        SendWebhookJob.perform_later('subscription.trial_ended', subscription)
       end
     end
 
     private
 
     attr_reader :timestamp
 
+    # This is to avoid billing at the end of the trial if the customer was billed at the beginning
+    # It's only for users who started billing customer AND upgraded their lago with this feature
+    # during the customer trial period
+    # Unfortunately, this introduces an N+1 query
+    def already_billed_on_day_one?(subscription)
+      subscription.invoice_subscriptions.where(timestamp: subscription.started_at.all_day).exists?
+    end
+
     def ending_trial_subscriptions
       sql = <<-SQL
         WITH
-          initial_started_at AS (#{initial_started_at})
+          initial_started_at AS (#{initial_started_at}),
+          already_billed_today AS (#{already_billed_today})
         SELECT DISTINCT
           plans.pay_in_advance AS plan_pay_in_advance,
+          already_billed_today.invoiced_count > 0 AS was_already_billed_today,
+          #{trial_end_date} as trial_end_utc_date_from_query,
           subscriptions.*
         FROM
           subscriptions
           INNER JOIN plans ON subscriptions.plan_id = plans.id
           INNER JOIN initial_started_at ON initial_started_at.external_id = subscriptions.external_id
           INNER JOIN customers ON subscriptions.customer_id = customers.id
           INNER JOIN organizations ON customers.organization_id = organizations.id
+          LEFT JOIN already_billed_today ON already_billed_today.subscription_id = subscriptions.id
         WHERE
           subscriptions.status = 1
           AND subscriptions.trial_ended_at IS NULL
-          AND DATE(initial_started_at#{at_time_zone} + plans.trial_period * INTERVAL '1 day') = DATE('#{timestamp}'#{at_time_zone})
+          AND #{trial_end_date + at_time_zone} <= '#{timestamp}'#{at_time_zone}
+          AND #{trial_end_date + at_time_zone} > '#{timestamp - 2.hours}'#{at_time_zone}
```

Comment:
Maybe we should get rid of this. Any subsciptions that terminated and doesn't have `trial_ended_at` should be treated.